### PR TITLE
[Truffle] Adding annotation to return an enumerator if no block is given to a method.

### PR DESCRIPTION
**Pull request by bjfish**

### CODE REVIEW

File: truffle/src/main/java/org/jruby/truffle/nodes/core/ReturnEnumeratorIfNoBlockNode.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        this.method = method;
+        this.methodName = methodName;
+    }
+
+    @Override
+    public Object execute(VirtualFrame frame) {
+        final RubyProc block = RubyArguments.getBlock(frame.getArguments());
+
+        if (block == null) {
+
```
Comment (chrisseaton) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/ReturnEnumeratorIfNoBlockNode.java]: Sorry to be a pain, but we don't want a newline here...

File: truffle/src/main/java/org/jruby/truffle/nodes/core/ReturnEnumeratorIfNoBlockNode.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            if (toEnumNode == null) {
+                CompilerDirectives.transferToInterpreter();
+                toEnumNode = insert(DispatchHeadNodeFactory.createMethodCall(getContext()));
+            }
```
Comment (chrisseaton) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/ReturnEnumeratorIfNoBlockNode.java]: ...and we do want a blank line here!

File: truffle/src/main/java/org/jruby/truffle/nodes/core/ReturnEnumeratorIfNoBlockNode.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            return toEnumNode.call(frame, RubyArguments.getSelf(frame.getArguments()), "to_enum", null, getContext().getSymbolTable().getSymbol(methodName));
+
+        } else {
+
```
Comment (chrisseaton) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/ReturnEnumeratorIfNoBlockNode.java]: same

File: truffle/src/main/java/org/jruby/truffle/nodes/core/ReturnEnumeratorIfNoBlockNode.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            try {
+                return method.executeRubyBasicObject(frame);
+            } catch (UnexpectedResultException e) {
+                return e.getResult();
+            }
+
```
Comment (chrisseaton) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/ReturnEnumeratorIfNoBlockNode.java]: same

File: truffle/src/main/java/org/jruby/truffle/nodes/core/CoreMethodNodeManager.java [hunk 5]
```diff
@@ -210,6 +210,10 @@ private static RubyRootNode makeGenericMethod(RubyContext context, MethodDetails
         final CheckArityNode checkArity = new CheckArityNode(context, sourceSection, arity);
         RubyNode sequence = SequenceNode.sequence(context, sourceSection, checkArity, methodNode);
 
+        if (methodDetails.getMethodAnnotation().returnsEnumeratorIfNoBlock()) {
+            sequence = new ReturnEnumeratorIfNoBlockNode(methodDetails.getMethodAnnotation().names()[0], sequence);
```
Comment (chrisseaton) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/CoreMethodNodeManager.java]: Using the first name here might not be correct. As it currently is, if you call `collect` without a block it will do `to_enum :map`, which isn't always going to be correct. We can handle that later, but can you put a comment in for now?

**Comment by bjfish** [comment]
@chrisseaton I updated but wasn't sure if i got the adding and removing lines correct

File: truffle/src/main/java/org/jruby/truffle/nodes/core/ReturnEnumeratorIfNoBlockNode.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        super(method.getContext(), method.getEncapsulatingSourceSection());
+        this.method = method;
+        this.methodName = methodName;
+    }
+
+    @Override
+    public Object execute(VirtualFrame frame) {
+        final RubyProc block = RubyArguments.getBlock(frame.getArguments());
+
+        if (block == null) {
```
Comment (nirvdrum) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/ReturnEnumeratorIfNoBlockNode.java]: I think we want a `ConditionProfile` here.  We've essentially collapsed two specializations into one to consolidate a common pattern.  But a given call site is apt to only use one form or the other.

**Comment by bjfish** [comment]
@chrisseaton it would be nice if the style things could be checked with something like checkstyle

Also, I have had trouble using raiseIfFrozen or the IsFrozenNode in a method after the method is to_enum'ed. Any ideas? I might try to debug this later on.

File: truffle/src/main/java/org/jruby/truffle/nodes/core/ReturnEnumeratorIfNoBlockNode.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+            if (toEnumNode == null) {
+                CompilerDirectives.transferToInterpreter();
+                toEnumNode = insert(DispatchHeadNodeFactory.createMethodCall(getContext()));
+            }
+
+            return toEnumNode.call(frame, RubyArguments.getSelf(frame.getArguments()), "to_enum", null, getContext().getSymbolTable().getSymbol(methodName));
+
+        } else {
+            try {
+                return method.executeRubyBasicObject(frame);
```
Comment (nirvdrum) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/ReturnEnumeratorIfNoBlockNode.java]: This should probably be `execute` instead of `executeRubyBasicObject`.  I'm not sure how often it happens (maybe never), but this wrapper won't work if the specialization returns a primitive.

Comment (chrisseaton) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/ReturnEnumeratorIfNoBlockNode.java]: Yes, sorry these notes are both correct and important but I already merged - can submit another PR?

Comment (bjfish) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/ReturnEnumeratorIfNoBlockNode.java]: I'll submit a new PR
